mltensor-explainer.md

```diff
@@ -0,0 +1,421 @@
+
+A privacy-conscious user wants to perform real-time selfie segmentation of a video feed on their local device.
+
+Currently, using WebNN for this task would require - for each frame - an expensive readback of `GPUBuffer` data to script, uploading the data to the ML context device (which may be the same GPU!), copying the result back to script, and then uploading the frame to be rendered back into a `GPUBuffer`. This is unlikely to be performed in real-time.
+
+An `MLTensor` may be imported into WebGPU, which in the best case provides zero-copy buffer sharing between the two APIs, and in all cases provides a synchronization mechanism between the respective WebNN and WebGPU [timelines](https://www.w3.org/TR/webgpu/#programming-model-timelines), avoiding the need for expensive synchronization via script.
+
+```js
+// Create a couple MLTensors to be used to facilitate WebGPU interop.
+const mlTensor1 = await mlContext.createTensor({..., usage: MLTensorUsage.WEBGPU_INTEROP});
+const mlTensor2 = await mlContext.createTensor({..., usage: MLTensorUsage.WEBGPU_INTEROP});
+
+const applyEffectToFrame = async () => {
+  const gpuVideoTexture = gpuDevice.importExternalTexture({source: video});
+
+  // Rent out the MLTensor to WebGPU.  
+  const tensorizedGpuBuffer = gpuDevice.importExternalBuffer(mlTensor1);
+
+  // Create a bind group for `gpuVideoTexture`, create a command encoder, etc.
+  // to "tensorize" `gpuVideoTexture` and store the result in `tensorizedGpuBuffer`
+  // ...
+
+  gpuDevice.queue.submit([tensorizationCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN.
+  tensorizedGpuBuffer.destroy();
+
+  // Perform some inference described by `graph` on the frame
+  // (e.g. selfie segmentation)
+  mlContext.dispatch(
+    graph,
+    /*inputs=*/{'input': mlTensor1},
+    /*outputs=*/{'output': mlTensor2},
+  );
+
+  // Rent the other MLTensor out to WebGPU.
+  const tensorizedGpuBufferAfterInference = gpuDevice.importExternalBuffer(mlTensor2);
+
+  // Create a bind group for `tensorizedGpuBufferAfterInference`,
+  // create a command encoder, etc to feed `tensorizedGpuBufferAfterInference`
+  // into a GPU shader which may blur the frame or replace background sections
+  // and then render the result
+  // ...
+
+  gpuDevice.queue.submit([texturizeAndRenderCommandEncoder.finish()]);
+
+  // Return the buffer to WebNN for the next frame.
+  tensorizedGpuBufferAfterInference.destroy();
+
+  // Call this method for each frame.
+  video.requestVideoFrameCallback(applyEffectToFrame);
+}
+```
+
+## Design Discussion
+
+### Timelines
+
+WebNN uses a programming model similar to [WebGPU's](https://www.w3.org/TR/webgpu/#programming-model), in that compute tasks are posted to a timeline - which I've referred to as an "ML context timeline" throughout this document - separate from the content timeline (i.e. "script"). See [the WebGPU documentation of timelines](https://gpuweb.github.io/gpuweb/#programming-model-timelines) for more details.
```

Comment:
Hmm we haven't actually specified WebNN's timelines yet (tracked in #529) so I'm hesitant to make such assertive statements for now

Reply:
I would say the example code already implies this and could be more clear:
```js
// Rent out the MLTensor to WebGPU.  
const tensorizedGpuBuffer = gpuDevice.importExternalBuffer(mlTensor1);
```

The `MLTensor` being created by the `MLContext` must be made available to the `gpuDevice` which can be the same device also specified to createContext(), so they must be on an equivalent timeline.

Reply:
FYI https://github.com/webmachinelearning/webnn/pull/754#discussion_r1746357792 proposes making `importExternalBuffer()` async, in part to decouple these timelines in the eyes of the spec

Reply:
The explainer now contains this note: https://github.com/webmachinelearning/webnn/pull/754/files#diff-0ee55cff3e7c8d2280b48fa37962505a398c7a18df2eb78e0497c76042054cc8R277

Is there more to discuss here or can we resolve this thread?

Reply:
> Importing and returning the `MLTensor` are each points of synchronization between the respective WebNN and WebGPU [timelines](https://www.w3.org/TR/webgpu/#programming-model-timelines).

I have a slight preference to call out the device-queue timeline (separate from content/script timeline) since that is where we **must** control the execution of queue commands between WebNN and WebGPU. I'll leave it up to you if you want to comment it here or punt that for the real spec.